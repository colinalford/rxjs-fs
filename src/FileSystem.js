import fsModule from 'fs';
import pathModule from 'path';
import { Observable } from 'rxjs';
import $$observable from 'symbol-observable';
import { isScheduler } from 'rxjs/util/isScheduler';

class FileSystem extends Observable {
    constructor(source) {
        if (!source || typeof source === 'function' || typeof source !== 'object') {
            super(source);
        } else if (typeof source[$$observable] === 'function') {
            super();
            this.source = source[$$observable]();
        }
    }
    lift(operator) {
        const observable = new FileSystem(this);
        observable.operator = operator;
        return observable;
    }
    /**
    * Asynchronously emits the contents of a directory.
    * @returns FileSystem<{ String path }>
    **/
    static dir(path = './', options) {
        return FileSystem.of({ path }).dir(options);
    }
    /**
    * Asynchronously emits the contents of a directory and all subdirectories.
    * @returns FileSystem<{ String path }>
    **/
    static expandDir(path = './', options) {
        return FileSystem
            .of({ path })
            .expand(({ path }) => FileSystem
                .dir(path, options));
    }
    /**
    * @returns FileSystem<String path, fs.Stats stats>
    **/
    stats() {
        return this.mergeMap(
            ({ path, stats }) => !stats &&
                FileSystem.stat(path) ||
                FileSystem.of(stats),
            ({ path, ...rest }, stats) => ({
                path, ...rest, stats
            })
        );
    }
    /**
    * @returns FileSystem<String path, fs.Stats fstats>
    **/
    fstats() {
        return this.mergeMap(
            ({ path, fstats }) => !fstats &&
                FileSystem.fstat(path) ||
                FileSystem.of(fstats),
            ({ path, ...rest }, fstats) => ({
                path, ...rest, fstats
            })
        );
    }
    /**
    * @returns FileSystem<String path, fs.Stats fstats>
    **/
    lstats() {
        return this.mergeMap(
            ({ path, lstats }) => !lstats &&
                FileSystem.lstat(path) ||
                FileSystem.of(lstats),
            ({ path, ...rest }, lstats) => ({
                path, ...rest, lstats
            })
        );
    }
    /**
    * @returns FileSystem<String path>
    **/
    dir(options) {
        return this
            .isDirectory()
            .mergeMap(({ path }) => FileSystem.readdir(path, options))
            .mergeMap(
                (paths) => paths,
                (paths, path) => ({ path })
            );
    }
    file(options) {
        return this
            .isFile()
            .mergeMap(
                ({ path }) => FileSystem.readFile(path, options),
                ({ path, ...rest }, file) => ({
                    path, ...rest, file
                })
            );
    }
    isExt(ext) {
        return this.stats().isFile().filter(({ path }) =>
            pathModule.extname(path) === ext
        );
    }
    isFile() {
        return this.stats().filter(({ stats }) => stats.isFile());
    }
    isFIFO() {
        return this.stats().filter(({ stats }) => stats.isFIFO());
    }
    isSocket() {
        return this.stats().filter(({ stats }) => stats.isSocket());
    }
    isDirectory() {
        return this.stats().filter(({ stats }) => stats.isDirectory());
    }
    isBlockDevice() {
        return this.stats().filter(({ stats }) => stats.isBlockDevice());
    }
    isSymbolicLink() {
        return this.lstats().filter(({ lstats }) => lstats.isSymbolicLink());
    }
    isCharacterDevice() {
        return this.stats().filter(({ stats }) => stats.isCharacterDevice());
    }
}

FileSystem = wrapObservableMethods(FileSystem, Observable);
FileSystem = wrapFSModuleMethods(FileSystem, fsModule);

export { FileSystem };
export default FileSystem;

function wrapObservableMethods(FileSystem, Observable) {
    function createStaticWrapper(staticMethodName) {
        return function(...args) {
            return new FileSystem(Observable[staticMethodName](...args));
        }
    }
    for (const staticMethodName in Observable) {
        FileSystem[staticMethodName] = createStaticWrapper(staticMethodName);
    }
    FileSystem.bindCallback = (...args) => (...args2) => new FileSystem(Observable.bindCallback(...args)(...args2));
    FileSystem.bindNodeCallback = (...args) => (...args2) => new FileSystem(Observable.bindNodeCallback(...args)(...args2));
    return FileSystem;
}

function wrapFSModuleMethods(FileSystem, fsModule) {
    function createFSWrapper(fsMethod) {
        function wrappedFSMethod(...args) {
            function wrappedFSMethodDefer() {
                const { args, toObservable } = wrappedFSMethodDefer;
                return toObservable(...args);
            }

            const scheduler = isScheduler(args[args.length - 1]) ? args.pop() : undefined;
            const projection = typeof args[args.length - 1] === 'function' ? args.pop() : undefined;

            wrappedFSMethodDefer.args = args;
            wrappedFSMethodDefer.toObservable = FileSystem.bindNodeCallback(
                wrappedFSMethod.fsMethod, projection, scheduler
            );

            return FileSystem.defer(wrappedFSMethodDefer);
        }

        wrappedFSMethod.fsMethod = fsMethod;

        return wrappedFSMethod;
    }

    return ([
        'access', 'appendFile', 'chmod', 'chown', 'close', 'exists', 'fchmod',
        'fchown', 'fdatasync', 'fstat', 'fsync', 'ftruncate', 'futimes', 'lchmod',
        'lchown', 'link', 'lstat', 'mkdir', 'mkdtemp', 'open', 'read', 'readdir',
        'readFile', 'readlink', 'realpath', 'rename', 'rmdir', 'stat', 'symlink',
        'truncate', 'unlink', 'utimes', 'write', 'writeFile'
    ].reduce((FileSystem, fsMethodName) => {
        FileSystem[fsMethodName] = createFSWrapper(fsModule[fsMethodName]);
        return FileSystem;
    }, FileSystem));
}
