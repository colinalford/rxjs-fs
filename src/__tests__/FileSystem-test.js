import path from 'path';
import FileSystem from '../FileSystem';
import { Observable, Scheduler } from 'rxjs';

const dirName = `${__dirname}/test-files`;
const filePath = `${dirName}/test-file`;
const fileContents = 'test-contents';

describe('FileSystem', () => {
    it('should wrap the static Observable methods', () => {
        for (const staticObservableMethod in Observable) {
            expect(FileSystem.hasOwnProperty(staticObservableMethod)).toBe(true);
        }
    });
    it('should create the test files directory', () => {
        return FileSystem
            .mkdir(dirName, () => ({ path: dirName }))
            .stats()
            .do(({ path, stats }) => {
                expect(path).toBe(dirName);
                expect(stats.isDirectory()).toBe(true);
            })
            .last()
            .toPromise();
    });
    it('should create a file in the test files directory', () => {
        return FileSystem
            .writeFile(
                `${filePath}-0`,
                `${fileContents}-0`,
                () => ({ path: `${filePath}-0` })
            )
            .file()
            .do(({ path, file, stats }) => {
                expect(path).toBe(`${filePath}-0`);
                expect(stats.isFile()).toBe(true);
                expect(file.toString()).toBe(`${fileContents}-0`);
            })
            .toPromise();
    });
    it('should remove a file in the test files directory', () => {
        return FileSystem
            .unlink(`${filePath}-0`, () => `${filePath}-0`)
            .mergeMap(
                (unlinkedFilePath) => FileSystem.exists(unlinkedFilePath, (x) => !!x),
                (unlinkedFilePath, exists) => ({ unlinkedFilePath, exists })
            )
            .do(({ unlinkedFilePath, exists }) => {
                expect(exists).toBe(false);
                expect(unlinkedFilePath).toBe(`${filePath}-0`);
            })
            .toPromise();
    });
    it('should remove the test files directory', () => {
        return FileSystem
            .rmdir(dirName, () => dirName)
            .mergeMap(
                (removedDirName) => FileSystem.exists(removedDirName, (x) => !!x),
                (removedDirName, exists) => ({ removedDirName, exists })
            )
            .do(({ removedDirName, exists }) => {
                expect(exists).toBe(false);
                expect(removedDirName).toBe(dirName);
            })
            .toPromise();
    });
});
