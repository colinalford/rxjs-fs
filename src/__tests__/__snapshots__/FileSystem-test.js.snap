exports[`FileSystem should create many files in the test files directory 1`] = `"/Users/ptaylor/dev/rxjs-fs/src/__tests__/test-files/test-file-1"`;

exports[`FileSystem should create many files in the test files directory 2`] = `"test-contents-1"`;

exports[`FileSystem should create many files in the test files directory 3`] = `"/Users/ptaylor/dev/rxjs-fs/src/__tests__/test-files/test-file-2"`;

exports[`FileSystem should create many files in the test files directory 4`] = `"test-contents-2"`;

exports[`FileSystem should create many files in the test files directory 5`] = `"/Users/ptaylor/dev/rxjs-fs/src/__tests__/test-files/test-file-3"`;

exports[`FileSystem should create many files in the test files directory 6`] = `"test-contents-3"`;

exports[`FileSystem should create many files in the test files directory 7`] = `"/Users/ptaylor/dev/rxjs-fs/src/__tests__/test-files/test-file-4"`;

exports[`FileSystem should create many files in the test files directory 8`] = `"test-contents-4"`;

exports[`FileSystem should create many files in the test files directory 9`] = `"/Users/ptaylor/dev/rxjs-fs/src/__tests__/test-files/test-file-5"`;

exports[`FileSystem should create many files in the test files directory 10`] = `"test-contents-5"`;

exports[`FileSystem should create many files in the test files directory 11`] = `"/Users/ptaylor/dev/rxjs-fs/src/__tests__/test-files/test-file-6"`;

exports[`FileSystem should create many files in the test files directory 12`] = `"test-contents-6"`;

exports[`FileSystem should create many files in the test files directory 13`] = `"/Users/ptaylor/dev/rxjs-fs/src/__tests__/test-files/test-file-7"`;

exports[`FileSystem should create many files in the test files directory 14`] = `"test-contents-7"`;

exports[`FileSystem should create many files in the test files directory 15`] = `"/Users/ptaylor/dev/rxjs-fs/src/__tests__/test-files/test-file-8"`;

exports[`FileSystem should create many files in the test files directory 16`] = `"test-contents-8"`;

exports[`FileSystem should create many files in the test files directory 17`] = `"/Users/ptaylor/dev/rxjs-fs/src/__tests__/test-files/test-file-9"`;

exports[`FileSystem should create many files in the test files directory 18`] = `"test-contents-9"`;

exports[`FileSystem should create many files in the test files directory 19`] = `"/Users/ptaylor/dev/rxjs-fs/src/__tests__/test-files/test-file-10"`;

exports[`FileSystem should create many files in the test files directory 20`] = `"test-contents-10"`;

exports[`FileSystem should remove many files in the test files directory 1`] = `"/Users/ptaylor/dev/rxjs-fs/src/__tests__/test-files/test-file-1"`;

exports[`FileSystem should remove many files in the test files directory 2`] = `"/Users/ptaylor/dev/rxjs-fs/src/__tests__/test-files/test-file-2"`;

exports[`FileSystem should remove many files in the test files directory 3`] = `"/Users/ptaylor/dev/rxjs-fs/src/__tests__/test-files/test-file-3"`;

exports[`FileSystem should remove many files in the test files directory 4`] = `"/Users/ptaylor/dev/rxjs-fs/src/__tests__/test-files/test-file-4"`;

exports[`FileSystem should remove many files in the test files directory 5`] = `"/Users/ptaylor/dev/rxjs-fs/src/__tests__/test-files/test-file-5"`;

exports[`FileSystem should remove many files in the test files directory 6`] = `"/Users/ptaylor/dev/rxjs-fs/src/__tests__/test-files/test-file-6"`;

exports[`FileSystem should remove many files in the test files directory 7`] = `"/Users/ptaylor/dev/rxjs-fs/src/__tests__/test-files/test-file-7"`;

exports[`FileSystem should remove many files in the test files directory 8`] = `"/Users/ptaylor/dev/rxjs-fs/src/__tests__/test-files/test-file-8"`;

exports[`FileSystem should remove many files in the test files directory 9`] = `"/Users/ptaylor/dev/rxjs-fs/src/__tests__/test-files/test-file-9"`;

exports[`FileSystem should remove many files in the test files directory 10`] = `"/Users/ptaylor/dev/rxjs-fs/src/__tests__/test-files/test-file-10"`;
